### ISSUE TITLE
Sequence dictionaries should include SequenceRecords from second-in-pair reads

### ISSUE THREAD

**Issue by ryan-williams** [bigdatagenomics/adam#693]
I just hit an error:
```
htsjdk.samtools.SAMFormatException: Error parsing text SAM file. MRNM 'chr10' not found in any SQ record; Line 152
```
trying to load a SAM file, output by ADAM, that contains an `@SQ` record for the contig the read in question is on ("chr11"), but not that of its mapped mate ("chr10").

**Comment by ryan-williams** [comment]
Wow, I think I was just bit by the same underlying issue in a very differently-manifesting way.

tl;dr: why does [`SequenceRecord.fromADAMRecord` require reads to be the first of a pair](https://github.com/bigdatagenomics/adam/blob/2b5c8e959dd5c7a321dfb8d97b9c7b6b08a0eef6/adam-core/src/main/scala/org/bdgenomics/adam/models/SequenceDictionary.scala#L266)?

```
if ((!rec.getReadPaired || rec.getFirstOfPair) && (rec.getContig != null || rec.getMateContig != null)) {
  …
} else
  Set()
```

I was observing very strange behavior where `ShuffleRegionJoin` and `BroadcastRegionJoin` would return different results, even on trivial datasets (e.g. 1 read, 1 interval); the former would return an empty RDD.

At first I was omitting the `SequenceDictionary` argument to `ShuffleRegionJoin` (which from what I can tell should not be allowed; I'll file a separate issue for that), resulting in an empty `SequenceDictionary` and empty results. After catching that, I was constructing a `SequenceDictionary` from my `RDD[AlignmentRecord]`, but still getting empty results, apparently due to my `RDD` only having 2nd-of-pair reads in it.

I'm going to update this issue title to describe what I think the problem is.

**Comment by fnothaft** [comment]
Fascinating! I would say that's a bug in `SequenceDictionary`. The no-`SequenceDictionary` version of the `ShuffleRegionJoin` should compute the `SequenceDictionary` using the [ADAMSequenceDictionaryRDDAggregator](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMRDDFunctions.scala#L86) class. If the `SequenceDictionary` implementation was correct, your results should match.

It looks like the behavior in the `SequenceDictionary` came in via https://github.com/bigdatagenomics/adam/commit/6622e89168ba962b5b87ab955766a4d0d6e92967. @mlinderm any thoughts about this?

**Comment by ryan-williams** [comment]
Yea, so I also want to talk about the no-`SequenceDictionary` version of `ShuffleRegionJoin`; it seems like it currently always returns no intersections? Unless I'm mistaken, I really shot myself in the foot due to that (and this) for a few hours tonight, I think we should kill that overloaded version and [the `sd` singleton](https://github.com/bigdatagenomics/adam/blob/2b5c8e959dd5c7a321dfb8d97b9c7b6b08a0eef6/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala#L32).

It's not obvious to me how we would work a `ADAMSequenceDictionaryRDDAggregator` in there, either...

**Comment by fnothaft** [comment]
Ah, I see what you're saying. You're correct there. I think the correct way to do this might involve rethinking the `RegionJoin` trait and going from there.

**Comment by ryan-williams** [comment]
Partially answering my own question, you mean we should use [`ADAMSpecificRecordSequenceDictionaryRDDAggregator`](https://github.com/bigdatagenomics/adam/blob/2b5c8e959dd5c7a321dfb8d97b9c7b6b08a0eef6/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMRDDFunctions.scala#L136) to basically get the `SequenceDictionary`-relevant fields out of whichever object we are joining, by ~reflection?

That sounds bad. Maybe I'm missing something, but it seems like the magic required to auto-generate the SD inside SRJ would be too magical, and brittle.

**Comment by ryan-williams** [comment]
Ah, I just got what you meant by rethinking the `RegionJoin` trait: the no-SD API is inherited.

So, up until today I was looking at / using the `0.16.0` versions of the region-joins, where there is no pretense of a super-trait.

I think we could/should just go back to that. The use case strawman: "I have this `RegionJoin` that could be a `ShuffleRegionJoin` or a `BroadcastRegionJoin` and I don't know or care which one it is" doesn't seem to be that substantiated yet (please correct me if I'm wrong, I basically only paged in all of this stuff Saturday :) ).

The semantics/constraints of Shuffle vs. Broadcast are different enough that you kind of have to deliberately choose which one to use in each situation, so I don't think removing the super-trait is a very high cost, or even a cost at all.

**Comment by ryan-williams** [comment]
(suggest we port further discussion, if any, to #697)
